### PR TITLE
increase pod memory limit to 768

### PR DESCRIPTION
**Pull request by TomasTomecek**
because of npm:
https://github.com/packit/packit-service/issues/917#issuecomment-760948922

see the commit messages for details, I can provide more details if needed

### CODE REVIEW

File: sandcastle/api.py [hunk 1]
```diff
@@ -481,9 +481,13 @@ def _prepare_mdir_exec(
...
 
-        # git checkout - oc cp does not preserve the file mode and makes the repo dirty
         # set -e - fail the script when a command fails
-        script_template = f"#!/bin/bash\nset -e\ncd {unique_dir}\nexec {cmd_str}\n"
+        script_template = (
+            "#!/bin/bash\n"
+            "set -e\n"
+            "source /home/sandcastle/setup_env_in_openshift.sh\n"
+            f"cd {unique_dir}\nexec {cmd_str}\n"
+        )
```
Comment (csomh) [inline sandcastle/api.py:498]: Why was this change made? Please take a note of this in the commit message.

File: sandcastle/api.py [hunk 2]
```diff
@@ -185,6 +185,10 @@ def create_pod_manifest(self, command: Optional[List] = None) -> dict:
             "name": self.pod_name,
             "env": env_image_vars,
             "imagePullPolicy": "IfNotPresent",
+            # npm loves to make beautiful terminal output full of colors and sunshine
+            # which would look hideous in logs, so we should allocate a TTY
+            # unless this breaks some tool
```
Comment (csomh) [inline sandcastle/api.py:190]: Hm, I don't really get this explanation: I see `tty: True` commented out, but I don't understand why?

Comment (TomasTomecek) [reply sandcastle/api.py:190]: I'll explain further

File: sandcastle/api.py (hunk 1 above)
Comment (TomasTomecek) [reply sandcastle/api.py:498]: will add a test case for this and further explanation

Comment (TomasTomecek) [reply sandcastle/api.py]: I dropped that comment into a new PR since it started to be pretty big

**Review by jpopelka — APPROVED** [review]

File: sandcastle/api.py [hunk 3]
```diff
@@ -185,6 +185,14 @@ def create_pod_manifest(self, command: Optional[List] = None) -> dict:
...
             "env": env_image_vars,
             "imagePullPolicy": "IfNotPresent",
+            # we may be tempted to enable tty in the pod to get pretty terminal output
+            # but it's not as simple as that, for example
+            # npm loves to make beautiful terminal output full of colors and sunshine
+            # which would look hideous in ASCII logs (terminal escape sequences),
+            # therefore only allocate tty if you are debugging something locally by
+            # invoking real shell inside, tty should be set to false by default
+            # TODO: making this configurable would be the best
+            # "tty": True,
```
Comment (martinpitt) [inline sandcastle/api.py:195]: I just want to call out this: Thanks for being thorough here, and still make it easy to debug locally! :hugs:

**Comment by martinpitt** [comment]
Thanks for this! I just re-ran https://github.com/cockpit-project/cockpit-podman/pull/649 and it still [failed in the same way](https://prod.packit.dev/srpm-build/12229/logs). As I mentioned in https://github.com/packit/packit.dev/pull/226, I can now reproduce the failure with 512 and 768 MB of RAM, it only works with 1024 MB. BTW, it's not really npm that eats so much memory, it's the webpack step. That really slurps in tons of modules and does the heavy lifting to assemble the pages, unfortunately I don't know how to make this less hungry.. :sob:

**Comment by TomasTomecek** [comment]
That was one of the things I checked and it seems you can pass arguments to node to make it less memory hungry:

https://medium.com/the-node-js-collection/node-js-memory-management-in-container-environments-7eb8409a74e8

It's pity one cannot configure this via env vars or a config file :/

**Comment by martinpitt** [comment]
@TomasTomecek Thanks for digging this out! I'm experimenting with that.
```
$ git clean -ffdx; podman run -ti -e NODE_OPTIONS=--max-old-space-size=200 --rm --memory 768MB -v $PWD:/src -w /src docker.io/usercont/sandcastle:prod make dist-gzip

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x7f66c3a09a39]
Security context: 0x1421a361b7b9 <JSObject>
    1: _serializeMappings(aka SourceMapGenerator_serializeMappings) [0x3b221bab7c01] [/src/node_modules/webpack-sources/node_modules/source-map/lib/source-map-generator.js:~312] [pc=0x1885e640e821](this=0x286358800611 <SourceMapGenerator map = 0xbe3cae9e3e9>)
    2: toJSON(aka SourceMapGenerator_toJSON) [0x3b221bab7c81] [/src/node_modules/webpack-sources/nod...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x7f66c2b995cc node::Abort() [/lib64/libnode.so.72]
 2: 0x7f66c29d926c node::OnFatalError(char const*, char const*) [/lib64/libnode.so.72]
 3: 0x7f66c2f0bb5a v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/lib64/libnode.so.72]
 4: 0x7f66c2f0bde2 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/lib64/libnode.so.72]
 5: 0x7f66c308bf99  [/lib64/libnode.so.72]
 6: 0x7f66c30a0927 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/lib64/libnode.so.72]
 7: 0x7f66c30a168a v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/lib64/libnode.so.72]
 8: 0x7f66c30a19f0 v8::internal::Heap::CollectAllGarbage(int, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/lib64/libnode.so.72]
 9: 0x7f66c30566a2 v8::internal::StackGuard::HandleInterrupts() [/lib64/libnode.so.72]
10: 0x7f66c337a855 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/lib64/libnode.so.72]
11: 0x7f66c3a09a39  [/lib64/libnode.so.72]
make: *** [Makefile:75: dist/manifest.json] Aborted (core dumped)
```

So 200 MB "old space" is too little, but this definitively did something. That suggests that it can be set through `$NODE_OPTIONS`, and thus I could add that to our packit.yml (unless you prefer adding it to sandcastle globally).

```
git clean -ffdx; podman run -ti -e NODE_OPTIONS=--max-old-space-size=400 --rm --memory 768MB -v $PWD:/src -w /src docker.io/usercont/sandcastle:prod make dist-gzip
```
This worked!

I'm trying to change the parameters to maybe squeeze it into 512 MB even, as that PR apparently isn't rolled out to production yet.

**Comment by TomasTomecek** [comment]
no worries and thinks for such a detailed investigation!

we already purchased more memory for our cluster and can increase our sandbox pods to 768M :)

**Pull request merged**